src/linear_algebra/pi_tensor_product.lean

```diff
@@ -0,0 +1,427 @@
+* API for the various ways `ι` can be split into subsets; connect this with the binary
+  tensor product.
+* Include connection with holors.
+* Port more of the API from the binary tensor product over to this case.
+
+## Tags
+
+multilinear, tensor, tensor product
+-/
+
+noncomputable theory
+open_locale classical
+open function
+
+section semiring
+
+variables {ι : Type*} {R : Type*} [comm_semiring R]
+variables {s : ι → Type*} [∀ i, add_comm_monoid (s i)] [∀ i, semimodule R (s i)]
+variables {E : Type*} [add_comm_monoid E] [semimodule R E]
+variables {F : Type*} [add_comm_monoid F]
+
+namespace pi_tensor_product
+include R
+variables (R) (s)
+
+/-- The relation on `free_add_monoid (R × Π i, s i)` that generates a congruence whose quotient is
+the tensor product. -/
+inductive eqv : free_add_monoid (R × Π i, s i) → free_add_monoid (R × Π i, s i) → Prop
+| of_zero : ∀ (r : R) (f : Π i, s i) (i : ι) (hf : f i = 0), eqv (free_add_monoid.of (r, f)) 0
+| of_zero_scalar : ∀ (f : Π i, s i), eqv (free_add_monoid.of (0, f)) 0
+| of_add : ∀ (r : R) (f : Π i, s i) (i : ι) (m₁ m₂ : s i), eqv
+    (free_add_monoid.of (r, update f i m₁) + free_add_monoid.of (r, update f i m₂))
+    (free_add_monoid.of (r, update f i (m₁ + m₂)))
+| of_add_scalar : ∀ (r r' : R) (f : Π i, s i), eqv
+    (free_add_monoid.of (r, f) + free_add_monoid.of (r', f))
+    (free_add_monoid.of (r + r', f))
+| of_smul : ∀ (r : R) (f : Π i, s i) (i : ι) (r' : R), eqv
+    (free_add_monoid.of (r, update f i (r' • (f i))))
+    (free_add_monoid.of (r' * r, f))
+| add_comm : ∀ x y, eqv (x + y) (y + x)
+
+end pi_tensor_product
+
+variables (R) (s)
+
+/-- `pi_tensor_product R s` with `R` a commutative semiring and `s : ι → Type*` is the tensor
+  product of all the `s i`'s. This is denoted by `⨂[R] i, s i`. -/
+def pi_tensor_product : Type* :=
+(add_con_gen (pi_tensor_product.eqv R s)).quotient
+
+variables {R}
+
+/- This enables the notation `⨂[R] i : ι, s i` for the pi tensor product, given `s : ι → Type*`. -/
+localized "notation `⨂[`:100 R `] ` binders `, ` r:(scoped:67 f, pi_tensor_product R f) := r"
+  in tensor_product
+
+open_locale tensor_product
+
+namespace pi_tensor_product
+
+section module
+
+instance : add_comm_monoid (⨂[R] i, s i) :=
+{ add_comm := λ x y, add_con.induction_on₂ x y $ λ x y, quotient.sound' $
+    add_con_gen.rel.of _ _ $ eqv.add_comm _ _,
+  .. (add_con_gen (pi_tensor_product.eqv R s)).add_monoid }
+
+instance : inhabited (⨂[R] i, s i) := ⟨0⟩
+
+variables (R) {s}
+
+/-- `tprod_coeff R r f` with `r : R` and `f : Π i, s i` is the tensor product of the vectors `f i`
+over all `i : ι`, multiplied by the coefficient `r`. -/
+def tprod_coeff (r : R) (f : Π i, s i) : ⨂[R] i, s i := add_con.mk' _ $ free_add_monoid.of (r, f)
+
+/-- `tprod R f` with `f : Π i, s i` is the tensor product of the vectors `f i` over all `i : ι`.
+
+This is bundled into a `multilinear_map` in `pi_tensor_product.mk`. -/
+def tprod (f : Π i, s i) : ⨂[R] i, s i := tprod_coeff R 1 f
```

Comment:
I wonder if it's worth eliminating `tprod_coeff` entirely, and just using
```suggestion
/-- `tprod R f` with `f : Π i, s i` is the tensor product of the vectors `f i` over all `i : ι`.

This is bundled into a `multilinear_map` in `pi_tensor_product.mk`. -/
def tprod (f : Π i, s i) : ⨂[R] i, s i := add_con.mk' _ $ free_add_monoid.of (1, f)
```
Otherwise `tprod R r f` is just an extra function to have to write API about, that is a worse spelling of `r • tprod R f`.

Reply:
I don't think `tprod_coeff` should be used outside this file, but not having it makes it very awkward to prove a lot of the API here.

Reply:
I agree, `tprod_coeff` seems pretty essential to buildnig up the structure.